MPS: Connect to GameLift with Player Session Data (#370)
o3de/o3de-multiplayersample: Documentation/Hosting/AmazonGameLift/GameLift.md

Reviewer: If you call this file README.md, Github will auto display when users navigate to the folder.

Documentation/Hosting/AmazonGameLift/README.md

Author: Rearranged docs a bit. README in the gem itself. This is the natural place to look based on walk-the-engine feedback.

# MultiplayerSample Project with Amazon GameLift

> This README covers testing and running MultiplayerSample with Amazon GameLift.

Reviewer: Link to GameLift as a service, call out this is entirely optional and describe what GameLift does.

ie

```suggestion
This README covers optional setup, testing and running on [Amazon GameLift](https://aws.amazon.com/gamelift/), an AWS service to make hosting and scaling game servers easier. 
```

Author: Fixed

Author: Thank you! Added your changes

## Running with [Amazon GameLift](https://docs.aws.amazon.com/gamelift/index.html)

### Setup
1. Install the [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html), if you don't already have it. You will need it to interact with Amazon GameLift.

> 1. Enable the "AWSGameLift" and "MPSGameLift" gem by adding them to MultiplayerSample/Gem/Code/enabled_gems.cmake

Reviewer 2: ```suggestion
1. Enable the "AWSGameLift" and "MPSGameLift" gems by adding them to MultiplayerSample/Gem/Code/enabled_gems.cmake
```

> 1. Build the server, game launchers, and asset bundler for MultiplayerSample as normal, per [top-level README](/README.md).
>     ```sh
>     cmake --build build\windows --target Editor MultiplayerSample.GameLauncher MultiplayerSample.ServerLauncher AssetBundler --config profile -- /m /nologo
>     ```
> 1. Build all the assets

Reviewer 3: These lines need to end with two <spaces> to force a line break in markdown.

```suggestion
1. Build the server, game launchers, and asset bundler for MultiplayerSample as normal, per [top-level README](/README.md).  
    ```sh
    cmake --build build\windows --target Editor MultiplayerSample.GameLauncher MultiplayerSample.ServerLauncher AssetBundler --config profile -- /m /nologo
    ```
1. Build all the assets.  
```

Examining the markdown preview shows that all the steps are on a single line

Author: Did a markdown pass and found lots of errors. Fixed! Thanks!

    ```sh
    cmake --build build\windows --target MultiplayerSample.Assets --config profile -- /m /nologo
    ```

> 1. Work in progress (WiP) step: Add your AWS region to Config/default_aws_resource_mappings.json (example: "Region": "us-west-2")

Reviewer 2: is this step no longer WIP?

Author: Still WiP. "Configure" requires a region, but I havent investigated why.

    a. Currently needed otherwise when the client initializes GameLift there will be an error about not having a region. 
    b. This step will be removed once we properly parse the game-session data which contains the fleet-id, region-id, etc  

> ## Build Server for Windows (WiP)

Reviewer 2: is this no longer WIP?

Author: Still WiP. "Configure" requires a region, but I havent investigated why.

> 1. WiP Step: Build a profile pak server/game build

Reviewer: Link to the these docs: https://github.com/o3de/o3de-multiplayersample/blob/development/Documentation/PackedAssetBuilds.md

Where we can, we should avoid content duplication.

Reviewer 3: +1 on linking to the existing documents.

I do think it would be better however to use export-project python command that @tkothadev addedsupport to make it simpler for user to performs these steps.

Author: Spoke to @tkothadev, really cool stuff. I'll be implementing a script. Hoping to erase dozens of lines from this readme :)

    ```
    
    ```sh
    C:\GameLiftPackageWindows\MultiplayerSample.GameLauncher.exe -bg_ConnectToAssetProcessor=0 -sys_PakPriority=2 --connect
    ```

    Note: launch_server.cfg is required because there's a bug with multiplayer when calling --loadlevel in the command-line. See https://github.com/o3de/o3de/issues/15773.
1. Open C:\GameLiftPackageWindows\user\log\Server.log
    You should see the "New Starbase" level loaded
    ```
    Entities from new root spawnable 'levels/newstarbase/newstarbase.spawnable' are ready
    ```

## Prepare for GameLift
### Upload the build to GameLift
Note: Builds are tied to Fleets; you may want to delete the existing build and fleet via the AWS Gamelift dashboard just so you don't accidentally reference old builds or old fleets in future steps. 
`
aws gamelift upload-build --operating-system WINDOWS_2016 --build-root C:\GameLiftPackageWindows\ --name MultiplayerSample --build-version v1.0 --region us-west-2
`
Record BuildId for the next step. Example: build-fe4a2b2c-8ae5-4757-99a6-0ff372d03512

### Create Fleet
After running this command it'll take about an hour for the fleet to activate. Check the status on the GameLift dashboard. 

`

> aws gamelift create-fleet --region us-west-2 --name GameLiftO3DTest2016 --ec2-instance-type c5.large --fleet-type ON_DEMAND --build-id <BuildId> --runtime-configuration "GameSessionActivationTimeoutSeconds=300, MaxConcurrentGameSessionActivations=2, ServerProcesses=[{LaunchPath=C:\game\MultiplayerSample.ServerLauncher.exe, Parameters= --rhi=null -sys_PakPriority=2 -NullRenderer -sv_terminateOnPlayerExit=true -bg_ConnectToAssetProcessor=0 --sv_gameLiftEnabled=true --sv_dedicated_host_onstartup=false --console-command-file=launch_server.cfg, ConcurrentExecutions=1}]" --ec2-inbound-permissions "FromPort=33450,ToPort=34449,IpRange=0.0.0.0/0,Protocol=UDP"

Reviewer 3: You may want to [add a note using markdown](https://stackoverflow.com/a/43120795) for readers that the choice of fleet type has a real monetary cost the user should be aware of.

Nothing has to be added about the exact cost, just a reminder to the user that this will use AWS resources with a cost.

Author: good call. fixed
